persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java

```diff
@@ -158,6 +170,66 @@
     TYPE_MAPPINGS = ImmutableMap.copyOf(types);
   }
 
+  public static ByteBuffer toPagingState(PagingPosition pos, Parameters parameters) {
+    Row row = pos.currentRow();
+    Set<Pair<String, String>> tables =
+        row.columns().stream()
+            .map(c -> Pair.create(c.keyspace(), c.table()))
+            .collect(Collectors.toSet());
+
+    if (tables.isEmpty()) {
+      throw new IllegalArgumentException("Missing table information in custom paging request.");
+    }
+
+    if (tables.size() > 1) {
+      throw new IllegalArgumentException("Too many tables are referenced: " + tables);
+    }
+
+    Pair<String, String> table = tables.iterator().next();
+    CFMetaData cfm = Schema.instance.getCFMetaData(table.left, table.right);
+    if (cfm == null) {
+      throw new IllegalStateException("Table not found: " + table);
+    }
+
+    MultiCBuilder builder =
```

Comment:
Why not use a `CBuilder` here instead of `MultiCBuilder`?  That's from looking in the implementation of `ClusteringComparator#make()`.  If possible maybe it's worth using `ClusteringComparator#make()` instead:

```java
    List<ByteBuffer> partitionKey = new ArrayList<>(cfm.partitionKeyColumns().size());

    for (ColumnDefinition c : cfm.partitionKeyColumns()) {
      ByteBuffer value = row.getBytesUnsafe(c.name.toCQLString());

      if (value == null) {
        throw new IllegalArgumentException(
            String.format(
                "Partition key value is not present in current row (table: %s, column: %s)",
                table, c.name.toCQLString()));
      }
      partitionKey.add(value);
    }

    Clustering clustering = cfm.getKeyValidatorAsClusteringComparator().make(partitionKey.toArray());
```

It's simpler, but creates a bit more garbage.

Reply:
It looks like something like this possible for other backends too.

Reply:
Updated in 478d32619c74bb0b0b9c6dbd4d31c0445a9ff4b4 (force pushed)